Sources/FoundationEssentials/Data/Data+Writing.swift

```diff
@@ -387,44 +389,112 @@ private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPoint
 
-        do {
-            try write(buffer: buffer, toFileDescriptor: fd, path: inPath, parentProgress: callback)
-        } catch {
-            try auxPath.withNTPathRepresentation { pwszAuxPath in
-                _ = DeleteFileW(pwszAuxPath)
-            }
+    _close(fd)
+    fd = -1
 
-            if callback?.isCancelled ?? false {
-                throw CocoaError(.userCancelled)
-            } else {
-                throw CocoaError.errorWithFilePath(inPath, errno: errno, reading: false)
-            }
+    try auxPath.withNTPathRepresentation { pwszAuxiliaryPath in
+        var hFile = CreateFileW(pwszAuxiliaryPath, DELETE,
+                                FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                                nil, OPEN_EXISTING,
+                                FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT,
+                                nil)
+        if hFile == INVALID_HANDLE_VALUE {
+            let dwError = GetLastError()
+            _ = DeleteFileW(pwszAuxiliaryPath)
+            throw CocoaError.errorWithFilePath(inPath, win32: dwError, reading: false)
+        }
+        defer {
+            if hFile == INVALID_HANDLE_VALUE { return }
+            _ = CloseHandle(hFile)
         }
 
-        writeExtendedAttributes(fd: fd, attributes: attributes)
+        try inPath.path.withNTPathRepresentation { pwszPath in
+            let cbSize = wcslen(pwszPath) * MemoryLayout<WCHAR>.size
+            let dwSize = DWORD(MemoryLayout<FILE_RENAME_INFO>.size + cbSize)
+            withUnsafeTemporaryAllocation(byteCount: Int(dwSize),
+                                          alignment: MemoryLayout<FILE_RENAME_INFO>.alignment) { pBuffer in
+                var pInfo = pBuffer.baseAddress?.bindMemory(to: FILE_RENAME_INFO.self, capacity: 1)
+                pInfo?.pointee.Flags = FILE_RENAME_FLAG_POSIX_SEMANTICS | FILE_RENAME_FLAG_REPLACE_IF_EXISTS
+                pInfo?.pointee.RootDirectory = nil
+                pInfo?.pointee.FileNameLength = DWORD(cbSize)
+                withUnsafePointer(to: &pInfo?.pointee.FileName) { pFileName in
+                    wcscpy_s(pFileName, wcslen(pwszPath) + 1, pwszPath)
+                }
 
-        _close(fd)
-        fd = -1
+                if !SetFileInformationByHandle(hFile, FileRenameInfoEx, pInfo, dwSize) {
+                    let dwError = GetLastError()
+                    if dwError == ERROR_ACCESS_DENIED {
+                        let dwAttributes = GetFileAttributesW(pwszPath)
+                        if dwAttributes > 0,
+                                dwAttributes & (FILE_ATTRIBUTE_READONLY | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM) != 0 {
+                            // The target file is read-only, hidden or a system file. Remove those attributes and try again.
+                            if SetFileAttributesW(pwszPath, dwAttributes & ~(FILE_ATTRIBUTE_READONLY | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM)) {
```

Comment:
For my context, could you elaborate a bit more on why this behavior is here? It seems a bit counterintuitive to me that if the client tries to write to a file they don't have access to write to (because it's read-only for example) that we'd change the attributes of the file to make it writeable first. I think I'd expect that if that were the case, we should error and instead require that the client make this file writeable first. Is there a particular scenario we're trying to account for here in which the client perceives the location as writeable even if it isn't?

Reply:
The scenario in my mind is that one of these attributes might be in use as a temporary thing while another thread is performing the operation. I'm happy to drop it and see if it occurs in practice though. It would help simplify the code quite a bit.

Reply:
I'm dropping this for now, I think that we can get away with it.